gh-127011: Add __str__ and __repr__ to ConfigParser

Agent-Hellboy opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-127011 -->
* Issue: gh-127011
<!-- /gh-issue-number -->

StanFromIreland commented on Lib/configparser.py:
```diff
@@ -690,6 +691,7 @@ def __init__(self, defaults=None, dict_type=_default_dict,
             self._read_defaults(defaults)
         self._allow_unnamed_section = allow_unnamed_section
 
+
```
```suggestion
```

StanFromIreland commented on Lib/configparser.py:
```diff
@@ -1048,6 +1054,39 @@ def __iter__(self):
+        }
+        return f"<ConfigParser: {config_dict}>"
+
+
```
```suggestion
```
1 line is standard in this file.

StanFromIreland commented on Lib/configparser.py:
```diff
@@ -1048,6 +1054,39 @@ def __iter__(self):
+        return (f"<{self.__class__.__name__}("
+                f"params={init_params}, "
+                f"state={state_summary})>")
+
```
```suggestion
```

picnixz commented on Misc/ACKS:
```diff
@@ -1599,6 +1599,7 @@ Erik Rose
 Mark Roseman
 Josh Rosenberg
 Jim Roskind
+Prince Roshan
```
Can you add this entry one line above (h coming before k)

Agent-Hellboy replied:
Done

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,39 @@ def test_set_nonstring_types(self):
+        cf.set("sect2", "option2", "bar")
+
+
+        expected_str = (
```
```suggestion

        expected_str = (
```

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,39 @@ def test_set_nonstring_types(self):
+        )
+        self.assertEqual(str(cf), expected_str)
+
+
```
```suggestion
```

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,39 @@ def test_set_nonstring_types(self):
         self.assertRaises(TypeError, cf.set, "sect", 123, "invalid opt name!")
         self.assertRaises(TypeError, cf.add_section, 123)
 
+    def test_str_and_repr(self):
```
Let's have separate tests (one for `str`, one for `repr`)

Agent-Hellboy replied:
Done

picnixz commented on Lib/configparser.py:
```diff
@@ -1048,6 +1054,39 @@ def __iter__(self):
+        }
+
+        if len(self._sections) > 5:
+            state_summary["sections_truncated"] = f"...and {len(self._sections) - 5} more"
```
```suggestion
        sections_count = len(self._sections)
        state_summary = {
            "loaded_sources": self._loaded_sources,
            "sections_count": sections_count,
            "sections": list(self._sections)[:5],  # limit to 5 section names for readability
        }

        if sections_count > 5:
            state_summary["sections_truncated"] = f"...and {sections_count - 5} more"
```

picnixz commented on Lib/configparser.py:
```diff
@@ -752,6 +754,7 @@ def read(self, filenames, encoding=None):
             try:
                 with open(filename, encoding=encoding) as fp:
                     self._read(fp, filename)
+                    self._loaded_sources.append(filename)
```
An alternative way is to avoid doing this and do this at the end instead:

```py
self._loaded_sources.extend(read_ok)
return read_ok
```

Or sync what's been read if something fails, to do it after the call `read_ok.append()` instead so that the `filename` is normalized as well (not a pathlib.Path object)

picnixz commented on Lib/configparser.py:
```diff
@@ -773,11 +776,13 @@ def read_file(self, f, source=None):
         """Read configuration from a given string."""
         sfile = io.StringIO(string)
         self.read_file(sfile, source)
+        self._loaded_sources.append(source)
```
This will be duplicated because `read_file` already does `self._loaded_sources.append(source)`.

Agent-Hellboy replied:
Yes, now I have added this check inside unittest as well.

picnixz commented on Lib/configparser.py:
```diff
@@ -1048,6 +1054,39 @@ def __iter__(self):
 
+    def __str__(self):
+        config_dict = {
+            section: {key: value for key, value in self.items(section, raw=True)}
```
```suggestion
            section: dict(self.items(section, raw=True))
```

It should be enough right?

picnixz commented on Misc/NEWS.d/next/Library/2025-04-25-18-29-10.gh-issue-127011.Ipem5z.rst:
```diff
@@ -0,0 +1 @@
+The ``__str__`` and ``__repr__`` methods have been added to the :class:`configparser.RawConfigParser`
```
```suggestion
Added pretty rendering of :class:`configparser.RawConfigParser` objects
via :meth:`~object.__str__` and :meth:`~object.__repr__`.
```

StanFromIreland replied:
```suggestion
Implement :meth:`~object.__str__` and :meth:`~object.__repr__`
for :class:`configparser.RawConfigParser` objects
```
How about this more formal version?

Agent-Hellboy replied:
yeah , I am picking this.

picnixz commented on Misc/NEWS.d/next/Library/2025-04-25-18-29-10.gh-issue-127011.Ipem5z.rst:
```diff
@@ -0,0 +1,2 @@
+Implement :meth:`~object.__str__` and :meth:`~object.__repr__`
+for :class:`configparser.RawConfigParser` objects
```
```suggestion
Implement :meth:`~object.__str__` and :meth:`~object.__repr__`
for :class:`configparser.RawConfigParser` objects.
```

picnixz commented on Lib/configparser.py:
```diff
@@ -757,6 +758,7 @@ def read(self, filenames, encoding=None):
             if isinstance(filename, os.PathLike):
                 filename = os.fspath(filename)
             read_ok.append(filename)
+        self._loaded_sources.extend(read_ok)
```
```suggestion
            read_ok.append(filename)
            self._loaded_sources.append(filename)
```

Sorry this one is probably better because if we fail here for some reason, then we still want to say that we've read some sources (even if it failed) I think.

picnixz commented on Lib/configparser.py:
```diff
@@ -1048,6 +1052,38 @@ def __iter__(self):
+
+        return (f"<{self.__class__.__name__}("
+                f"params={init_params}, "
+                f"state={state_summary})>")
```
We could use `=` f-strings:

```py
return f"<{self.__class__.__name__}({params=}, {state=})>"
```

but this requires the variables to be named "params" and "state" instead.

Agent-Hellboy replied:
I can change if you want but `init_params` and `state_summary` are more descriptive for those who are reading the code.

picnixz replied:
Mmh, yes. Let's leave keep it that way (I'll let Jaraco have the final say)

jaraco replied:
My instinct is that if "params" is suitable for the repr, it's suitable for reading the code, and conversely, if "init_params" is more descriptive for the reader, it would also be more descriptive for the person evaluating the repr. In my opinion, "params" and "state" are better choices because their meaning is fairly clear or readily inferred from the context, and shorter, single-word names are preferred to longer ones, so my preference would be to use the simpler names. I don't feel strongly about it, though.

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,50 @@ def test_set_nonstring_types(self):
+        self.maxDiff = None
+        cf = self.config_class(allow_no_value=True, delimiters=('=',), strict=True)
+        cf.add_section("sect1")
+        cf.add_section("sect2")
```
No need for that many sections right? we can have a smaller test for `__str__` I think.

Agent-Hellboy replied:
Done

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,50 @@ def test_set_nonstring_types(self):
+
+        dict_type = type(cf._dict).__name__
+
+        expected_repr = (
```
This is very hard to read I think. How about doing something like this:

```py
expected = f"<{classname}({params=}, {state=})>"
```

and you create the `classname`, `params` and `state` strings yourself? (possibly with a `StringIO` or with `''.join(parts)`? as everything is inlined it's hard to read.

Agent-Hellboy replied:
okay, I have separated out these three.

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,51 @@ def test_set_nonstring_types(self):
+            'sections_truncated': '...and 1 more',
+        }
+        expected = f"<{classname}({params=}, {state=})>"
+        self.assertEqual(repr(cf), expected)
```
```suggestion
        params = {
            'dict_type': type(cf._dict).__name__,
            'allow_no_value': True,
            'delimiters': ('=',),
            'strict': True,
            'default_section': 'DEFAULT',
            'interpolation': 'BasicInterpolation',
        }
        state = {
            'loaded_sources': ['<string>'],
            'sections_count': 6,
            'sections': ['sect1', 'sect2', 'sect3', 'sect4', 'sect5'],
            'sections_truncated': '...and 1 more',
        }
        expected = f"<{type(cf).__name__}({params=}, {state=})>"
        self.assertEqual(repr(cf), expected)
```

Sorry, I thought the `expected` wouldn't fit on 80 chars, but it fits, so it should be fine like this! By the way, I used `type(x).__name__` instead of `x.__class__.__name__` but I don't know which one is the best. I usually prefer the `__class__` one but I got told once that sometimes `type(x)` is preferrable but I can't remember why. So up to you (considering you used `type(cf._dict)`, I suggest using `type(cf)` as well).

Agent-Hellboy replied:
cpython codebase uses type(cf).__name__ I grepped one instance, __class__ can be overridden, but this is test , it's okay. 
btw, I guess except library it's okay, I saw the use of type(obj).__name__ in tools.

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,51 @@ def test_set_nonstring_types(self):
+        cf.add_section("sect6")
+        cf.set("sect1", "option1", "foo")
+        cf.set("sect2", "option2", "bar")
+        cf.read_string("")
```
Is this to have `'loaded_sources'== ['<string'>]`? Maybe add a comment saying why we read this?

Agent-Hellboy replied:
Done

picnixz reviewed (commented):
It looks good but I'll let @jaraco have the final say

picnixz commented on Lib/test/test_configparser.py:
```diff
@@ -980,6 +980,50 @@ def test_set_nonstring_types(self):
+        cf.add_section("sect6")
+        cf.set("sect1", "option1", "foo")
+        cf.set("sect2", "option2", "bar")
+        cf.read_string("") # To trigger the loading of sources
```
```suggestion
        cf.read_string("")  # to trigger the loading of sources
```

auvipy reviewed (approved):
looks good so far

jaraco commented on Lib/configparser.py:
```diff
@@ -757,6 +758,7 @@ def read(self, filenames, encoding=None):
             if isinstance(filename, os.PathLike):
                 filename = os.fspath(filename)
             read_ok.append(filename)
+            self._loaded_sources.append(read_ok)
```
I don't love how `_loaded_sources` is populated from two different places. In this diff, it's easy to see the effect, but once it's merged, it'll be somewhat difficult to recognize the different touch points of this change.

I wonder if instead, this append operation should go in `_read`, so it's called in one place.

Agent-Hellboy committed:
gh-127011: Add __str__ and __repr__ to ConfigParser

- Implement __str__ method showing formatted configuration content
- Implement __repr__ method with constructor parameters and state summary
- Add comprehensive tests for both methods
- Update ACKS and add NEWS entry

Agent-Hellboy committed:
fix test failure
